### ISSUE TITLE
Using ArrayConnector with array of integers fails

### ISSUE THREAD

**Issue by bmerrison** [NeuralEnsemble/PyNN#405]
The example of ArrayConnector given in the documentation uses integer values (0 or 1) to indicate connections. However, in the current version of PyNN this gives an exception in some circumstances (seems to be when the size of the post-synaptic population is greater than the size of the pre-synaptic one).
**Example:**
```
pop_pre = Population(1, MN_Roberts2014CellType())
pop_post = Population(5, MN_Roberts2014CellType())

conn_array = np.array([[0, 0, 1, 0, 0]])
connector = ArrayConnector(conn_array)

projection = Projection(pop_pre, 
                        pop_post, 
                        connector, 
                        StaticSynapse(weight=1, delay=0), 
                        receptor_type='ampa')
```
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/usr/lib/python2.7/dist-packages/IPython/utils/py3compat.pyc in execfile(fname, *where)
    202             else:
    203                 filename = fname
--> 204             __builtin__.execfile(filename, *where)

/home/bmerrison/Tadpole/Code/ex_networks/pynn_bug.py in <module>()
     17                         connector,
     18                         StaticSynapse(weight=1, delay=0),
---> 19                         receptor_type='ampa')

/usr/local/lib/python2.7/dist-packages/pyNN/neuron/projections.py in __init__(self, presynaptic_population, postsynaptic_population, connector, synapse_type, source, receptor_type, space, label)
     40                                    space, label)
     41         self._connections = dict((index, defaultdict(list)) for index in self.post._mask_local.nonzero()[0])
---> 42         connector.connect(self)
     43         self._presynaptic_components = dict((index, {}) for index in 
     44                                             self.pre._mask_local.nonzero()[0])

/usr/local/lib/python2.7/dist-packages/pyNN/connectors.py in connect(self, projection)
    951         connection_map = LazyArray(self.array, projection.shape)
    952         print connection_map
--> 953         self._connect_with_map(projection, connection_map)
    954 
    955 

/usr/local/lib/python2.7/dist-packages/pyNN/connectors.py in _connect_with_map(self, projection, connection_map, distance_map)
    230         """
    231         logger.debug("Connecting %s using a connection map" % projection.label)
--> 232         self._standard_connect(projection, connection_map.by_column, distance_map)
    233 
    234 

/usr/local/lib/python2.7/dist-packages/pyNN/connectors.py in _standard_connect(self, projection, connection_map_generator, distance_map)
    212                     # Connect the neurons
    213                     #logger.debug("Connecting to %d from %s" % (col, source_mask))
--> 214                     projection._convergent_connect(source_mask, col, **connection_parameters)
    215                     if self.callback:
    216                         self.callback(count/projection.post.local_size)

/usr/local/lib/python2.7/dist-packages/pyNN/neuron/projections.py in _convergent_connect(self, presynaptic_indices, postsynaptic_index, **connection_parameters)
     96             #logger.debug("Connecting neuron #%s to neuron #%s with synapse type %s, receptor type %s, parameters %s", pre_idx, postsynaptic_index, self.synapse_type, self.receptor_type, parameters)
     97             self._connections[postsynaptic_index][pre_idx].append(
---> 98                 self.synapse_type.connection_type(self, pre_idx, postsynaptic_index, **parameters))
     99 
    100     def _configure_presynaptic_components(self):

/usr/local/lib/python2.7/dist-packages/pyNN/neuron/simulator.py in __init__(self, projection, pre, post, **parameters)
    338         self.presynaptic_index = pre
    339         self.postsynaptic_index = post
--> 340         self.presynaptic_cell = projection.pre[pre]
    341         self.postsynaptic_cell = projection.post[post]
    342         if "." in projection.receptor_type:

/usr/local/lib/python2.7/dist-packages/pyNN/common/populations.pyc in __getitem__(self, index)
    158         """
    159         if isinstance(index, (int, numpy.integer)):
--> 160             return self.all_cells[index]
    161         elif isinstance(index, (slice, list, numpy.ndarray)):
    162             return self._get_view(index)

IndexError: index out of bounds

```
**Workaround:** Use True and False instead of 1 and 0. Maybe it's just a case of updating the documentation?

**Comment by bmerrison** [comment]
In fact, using an array of integers doesn't work properly whatever the size of the pre/post populations. I used it to generate a set of connections within a population (pre_pop=post_pop) and even though there was no exception the connectivity wasn't correct (many connections were missing).

**Comment by apdavison** [comment]
The first error you mention seems to occur only when the pre-synaptic population is of size 1. With `pyNN.neuron` I can reproduce the error traceback above. With `pyNN.nest` I get:

```python
Traceback (most recent call last):
  File "tmp/issue405.py", line 16, in <module>
    receptor_type='excitatory')
  File "/home/docker/dev/PyNN/pyNN/nest/projections.py", line 60, in __init__
    connector.connect(self)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 950, in connect
    self._connect_with_map(projection, connection_map)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 230, in _connect_with_map
    self._standard_connect(projection, connection_map.by_column, distance_map)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 212, in _standard_connect
    projection._convergent_connect(source_mask, col, **connection_parameters)
  File "/home/docker/dev/PyNN/pyNN/nest/projections.py", line 136, in _convergent_connect
    presynaptic_cells = self.pre.all_cells[presynaptic_indices]
IndexError: index 1 is out of bounds for size 1
```

With `pyNN.brian` I get:

```python
Traceback (most recent call last):
  File "tmp/issue405.py", line 17, in <module>
    receptor_type='excitatory')
  File "/home/docker/dev/PyNN/pyNN/brian/projections.py", line 106, in __init__
    connector.connect(self)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 950, in connect
    self._connect_with_map(projection, connection_map)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 230, in _connect_with_map
    self._standard_connect(projection, connection_map.by_column, distance_map)
  File "/home/docker/dev/PyNN/pyNN/connectors.py", line 212, in _standard_connect
    projection._convergent_connect(source_mask, col, **connection_parameters)
  File "/home/docker/dev/PyNN/pyNN/brian/projections.py", line 166, in _convergent_connect
    self._brian_synapses[i_group][j_group][i, j] = True
  File "/home/docker/env/neurosci/local/lib/python2.7/site-packages/brian/synapses/synapses.py", line 616, in __setitem__
    % (pre_slice[-1], len(self.source)))
ValueError: Presynaptic index 1 greater than number of presynaptic neurons (1)
```

Increasing the presynaptic population size to 2 removes the error. A postsynaptic population size of 1 also seems to work.

**Comment by apdavison** [comment]
On investigating further, it turns out this is a documentation problem, as you suggested.

The array definition should be
```python
conn_array = np.array([[0, 0, 1, 0, 0]], dtype=bool)
```
not just
```python
conn_array = np.array([[0, 0, 1, 0, 0]])
```
(writing explicit `True` and `False` in the array also works, of course, but is more verbose).

I've pushed a documentation fix: 384181ab2fe3a28bfb4293073d4f5d1afe4bb403
